uqcsbot/starboard.py

```diff
@@ -1,244 +1,598 @@
         db_session.close()
-    
-    def _create_sb_embed(self, recv: discord.Message) -> discord.Embed:
-        embed = discord.Embed(color=recv.author.top_role.color, description=recv.content)
-        embed.set_author(name=recv.author.display_name, icon_url=recv.author.display_avatar.url)
-        embed.set_footer(text=recv.created_at.strftime('%b %d, %H:%M:%S'))
-
-        if len(recv.attachments) > 0:
-            embed.set_image(url = recv.attachments[0].url)
+
+    async def _fetch_message_or_none(
+        self, channel: (discord.TextChannel | None), id: (int | None)
+    ) -> (discord.Message | None):
+        """Translates a lot of None into much less None. Also translates NotFound exceptions."""
+        if channel is None or id is None:
+            return None
+        else:
+            try:
+                return await channel.fetch_message(id)
+            except discord.NotFound():
+                return None
+
+    async def _lookup_from_id(
+        self, channel_id: int, message_id: int
+    ) -> Tuple[discord.Message | None, discord.Message | None]:
+        """Takes a channel ID and a message ID. These may be for a starboard message or a recieved message.
+
+        Returns (Recieved Message, Starboard Message), or Nones, as applicable.
+        """
+        db_session = self.bot.create_db_session()
+
+        entry = None
+        if self.starboard_channel.id == channel_id:
+            # we're primarily looking up a recieved message and a location.
+            # first, get the entry, then the location, then _fetch_message_or_none the remaining IDs.
+            entry = (
+                db_session.query(models.Starboard)
+                .filter(models.Starboard.sent == message_id)
+                .one_or_none()
+            )
+
+            if entry is not None:
+                if entry.recv_location is not None:
+                    channel = self.bot.get_channel(entry.recv_location)
+
+                    return (
+                        await self._fetch_message_or_none(channel, entry.recv),
+                        await self._fetch_message_or_none(
+                            self.starboard_channel, entry.sent
+                        ),
+                    )
+                else:
+                    # if the recieved location is None, then the message won't exist either.
+                    # The only thing we can possibly return is a starboard message.
+                    return (
+                        None,
+                        await self._fetch_message_or_none(
+                            self.starboard_channel, entry.sent
+                        ),
+                    )
+            else:
+                raise SomethingsFucked(
+                    modlog=self.modlog,
+                    message=f"Couldn't find an DB entry for this starboard message ({message_id})!",
+                    # note that this will also trigger on Isaac's initial-starboard message (1076779482637144105).
```

Comment:
Maybe also note this at the definition of the error.

Reply:
TBQH in the interests of reducing alert fatigue etc etc I'm just going to add an ID check there.

We want to avoid IDs wherever possible, so the bot behaves consistently across different servers. But this isn't going to cause errors if we run the bot in another server, and it's the best way to check genuine equality of two messages.

Reply:
I meant showing that the `SomethingsFucked` error should have a note that it will be called each time on the main server. I can see someone trying to debug this and not knowing why we are getting a  `SomethingsFucked` error.